### PR TITLE
[ADD] sale_commission_product_criteria_domain

### PR DESCRIPTION
**Pull request by ilyasProgrammer**

### CODE REVIEW

**Comment by francesco-ooops** [comment]
@ilyasProgrammer can you rebase?

**Review by francesco-ooops — APPROVED** [review]
Functional ok!

File: sale_commission_product_criteria_domain/models/commission.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        related="commission_id.commission_type", readonly=True
+    )
+    group_id = fields.Many2one(
+        "commission.items.group",
+        ondelete="restrict",
+    )
+
+    def write(self, values):
+        res = super().write(values)
+        if self.group_id and len(self.group_id.commission_ids.ids) == 0:
```
Comment (geomer198) [inline sale_commission_product_criteria_domain/models/commission.py]: Better change it on this:
```
if self.group_id and not self.group_id.commission_ids:
```

File: sale_commission_product_criteria_domain/models/commission.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+            self.group_id.commission_ids = [(6, 0, self.commission_id.ids)]
```
Comment (geomer198) [inline sale_commission_product_criteria_domain/models/commission.py:35]: Better use `write` function.

File: sale_commission_product_criteria_domain/models/commission.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+            partner = self.object_id.partner_id
+        elif self.object_id._name == "sale.order.line":
+            partner = self.object_id.order_id.partner_id
+        # if not partner or not partner.apply_commission_restrictions:
+        if not partner:
+            return super()._get_single_commission_amount(
+                commission, subtotal, product, quantity
+            )
+        item_ids = self._get_commission_items(commission, product)
+        if len(item_ids) == 0:
```
Comment (geomer198) [inline sale_commission_product_criteria_domain/models/commission.py]: If you want check is list empty you just use: `if not item_ids:`

File: sale_commission_product_criteria_domain/models/commission.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        # Main idea is here:
+        group_ids = partner.commission_item_agent_ids.filtered(
+            lambda x: x.agent_id == self.agent_id
+        ).mapped("group_ids")
+        ci_with_group_ids = self.env["commission.item"].search(
+            [
+                ("group_id", "in", group_ids.ids),
+            ]
+        )
+        item_ids = [item for item in item_ids if item in ci_with_group_ids.ids]
```
Comment (geomer198) [inline sale_commission_product_criteria_domain/models/commission.py]: Could you use intersection for lists as like this: `list(set(list1) & set(list2))`

File: sale_commission_product_criteria_domain/models/commission.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        if len(item_ids) == 0:
```
Comment (geomer198) [inline sale_commission_product_criteria_domain/models/commission.py]: Better this solution:
`if not item_ids:`

File: sale_commission_product_criteria_domain/models/commission.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            return 0.0
+        group_ids = partner.commission_item_agent_ids.filtered(
+            lambda x: x.agent_id == self.agent_id
+        ).mapped("group_ids")
+        ci_with_group_ids = self.env["commission.item"].search(
+            [
+                ("group_id", "in", group_ids.ids),
+            ]
+        )
+        item_ids = [item for item in item_ids if item in ci_with_group_ids.ids]
```
Comment (geomer198) [inline sale_commission_product_criteria_domain/models/commission.py]: Better use intersection for lists

File: sale_commission_product_criteria_domain/models/commission.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        if len(item_ids) == 0:
```
Comment (geomer198) [inline sale_commission_product_criteria_domain/models/commission.py]: ` if not item_ids:`

File: sale_commission_product_criteria_domain/models/commission_group.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    )
+    agents_count = fields.Integer(compute="_compute_agents_count")
+
+    @api.depends("item_ids")
+    def _compute_commission_ids(self):
+        for rec in self:
+            rec.commission_ids = [(6, 0, rec.item_ids.mapped("commission_id").ids)]
+
+    def unlink(self):
+        if len(self.item_ids) > 0:
```
Comment (geomer198) [inline sale_commission_product_criteria_domain/models/commission_group.py]:
```
if self.item_ids:
    ...
```

File: sale_commission_product_criteria_domain/models/commission_group.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+                _(
+                    "You can not delete this commission group since "
+                    "there is related to it commission items."
+                )
+            )
+        return super().unlink()
+
+    def _compute_agents_count(self):
+        for rec in self:
+            agent_ids = self.env["res.partner"].search(
```
Comment (geomer198) [inline sale_commission_product_criteria_domain/models/commission_group.py]: Take out `self.env["res.partner"] from cycle.
Just set variable `res_partner_obj = self.env["res.partner"] before cycle.
And use `search_count()` for get records count and set result in record field.
`self.agents_count = res_partner_obj.search_count(...)`

File: sale_commission_product_criteria_domain/models/partner.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+                {"partner_id": rec._origin.id, "agent_id": x._origin.id}
+                for x in rec.agent_ids.filtered(
+                    lambda x: x.commission_id.commission_type == "product_restricted"
+                )
+                if x not in exiting_agents.ids
+            ]
+            to_delete = rec.commission_item_agent_ids.filtered(
+                lambda x: x.agent_id.id in (exiting_agents - rec.agent_ids).ids
+            )
+            for line in to_create:
```
Comment (geomer198) [inline sale_commission_product_criteria_domain/models/partner.py]: move cycle into `update()`: `rec.update({"commission_item_agent_ids": [(0,0,line) for line in to_create])`

File: sale_commission_product_criteria_domain/models/partner.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+                rec.update({"commission_item_agent_ids": [(0, 0, line)]})
+            for del_line in to_delete:
```
Comment (geomer198) [inline sale_commission_product_criteria_domain/models/partner.py]: here too

**Review by geomer198 — COMMENTED** [review]
Please fix comments

File: sale_commission_product_criteria_domain/models/commission.py (hunk 2 above)
Comment (ilyasProgrammer) [reply sale_commission_product_criteria_domain/models/commission.py:35]: @geomer198 why ?

**Review by francesco-ooops — APPROVED** [review]
Functional ok after adding tests

As per description: 

This module allows to manage the following commission structure use case:

Agent A receives commission 50 when selling product 1 to customer X
Agent A receives commission 20 when selling product 1 to customer Y

File: sale_commission_product_criteria_domain/models/commission.py (hunk 2 above)
Comment (geomer198) [reply sale_commission_product_criteria_domain/models/commission.py:35]: I think this solution is possible.

**Review by geomer198 — APPROVED** [review]
Code Review ok!

**Comment by francesco-ooops** [comment]
@pedrobaeza can we merge? We worked hard on this :)

File: sale_commission_product_criteria_domain/__manifest__.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    "name": "Sale Commission Product Criteria Domain",
+    "version": "14.0.1.0.0",
+    "author": "Ilyas," "Ooops404," "Odoo Community Association (OCA)",
+    "contributors": ["Ilyas"],
+    "maintainers": ["ilyasProgrammer"],
+    "website": "https://github.com/OCA/commission",
+    "category": "Sales Management",
+    "license": "AGPL-3",
+    "depends": [
+        "sale_commission_product_criteria_discount",
```
Comment (pedrobaeza) [inline sale_commission_product_criteria_domain/__manifest__.py]: This dependency shows a non-desired coupling between modules. Maybe this should be split into 2: one for base product criteria, and another for the discounts.

Comment (ilyasProgrammer) [reply sale_commission_product_criteria_domain/__manifest__.py]: Thats right. I redid it to depend only on sale_commission_product_criteria

**Comment by ilyasProgrammer** [comment]
@pedrobaeza Could you merge please

File: sale_commission_product_criteria_domain/readme/DESCRIPTION.rst [hunk 13]
```diff
@@ -0,0 +1,7 @@
+This module allows to manage the following commission structure use case:
+
+Agent A receives commission 50 when selling product 1 to customer X
+
+Agent A receives commission 20 when selling product 1 to customer Y
+
+This behavior is integrated with both pricelist-like Commission Type structure (sale_commission_product_criteria) and commission by discount (sale_commission_product_criteria_discount)
```
Comment (pedrobaeza) [inline sale_commission_product_criteria_domain/readme/DESCRIPTION.rst]: Incorrect text?

**Pull request merged**